# Outlier checking

We focus on outliers beyond z score of 5 and examine by each feature carefully.

1 in 21613 samples ~0.005% ~ z score of 4, so using absolute value of z score of 5 as threshold for checking should be reasonable for first screening

Since we cannot double check data accruacy directly with sellers or buyers, we will only remove outliers conservatively and give the benefit of doubt to data.

After checking each feature carefully, we will provide a summary and decision at the end.

In [79]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats

path = './archive/'

kc_data = pd.read_csv(path + '/kc_house_data.csv')


In [80]:
# Droping 'id' and 'date' column since the first one is note a feature and the second one is not a numerical feature

data = kc_data.drop(['id','date'],axis = 1)
columns =[]
for col in data.columns:
    columns.append(col)
# Calculating z-score value
z_score = np.abs(stats.zscore(data))
z_score_df = pd.DataFrame(z_score, columns = columns)

outlier = data[(z_score_df > 5).all(axis=1)]
outlier

print("Examine outliers by each feature in the below:")

Examine outliers by each feature in the below:


In [81]:
data[z_score_df.price>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
246    2400000.0         4       2.50         3650      8354     1.0   
269    2900000.0         4       3.25         5050     20100     1.5   
300    3075000.0         4       5.00         4550     18641     1.0   
312    2384000.0         5       2.50         3650      9050     2.0   
656    3070000.0         3       2.50         3930     55867     1.0   
...          ...       ...        ...          ...       ...     ...   
20460  3345000.0         5       3.75         5350     15360     1.0   
20535  2950000.0         4       4.25         4470      5884     2.0   
21040  2900000.0         5       4.00         5190     14600     2.0   
21530  3000000.0         4       3.75         5090     14823     1.0   
21576  3567000.0         5       4.50         4850     10584     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
246             1     4          3      9        1830           1820   
269             0     2          3     11        4750            300   
300             1     4          3     10        2600           1950   
312             0     4          5     10        3370            280   
656             1     4          4      8        2330           1600   
...           ...   ...        ...    ...         ...            ...   
20460           0     1          3     11        3040           2310   
20535           0     1          3     11        3230           1240   
21040           0     1          3     11        5190              0   
21530           0     0          3     11        4180            910   
21576           1     4          3     10        3540           1310   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
246        2000             0    98074  47.6338 -122.072           3120   
269        1982          2008    98004  47.6312 -122.223           3890   
300        2002             0    98074  47.6053 -122.077           4550   
312        1921             0    98119  47.6345 -122.367           2880   
656        1957             0    98034  47.7022 -122.224           2730   
...         ...           ...      ...      ...      ...            ...   
20460      2008             0    98004  47.6480 -122.218           3740   
20535      2010             0    98199  47.6387 -122.405           2570   
21040      2013             0    98039  47.6102 -122.225           3840   
21530      2013             0    98004  47.6200 -122.207           3030   
21576      2007             0    98008  47.5943 -122.110           3470   

       sqft_lot15  
246         18841  
269         20060  
300         19508  
312          5400  
656         26324  
...           ...  
20460       15940  
20535        6000  
21040       19250  
21530       12752  
21576       18270  

[126 rows x 19 columns]

In [82]:
data[z_score_df.price>10]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
1164  5110800.0         5       5.25         8010     45517     2.0   
1315  5300000.0         6       6.00         7390     24829     2.0   
1448  5350000.0         5       5.00         8000     23985     2.0   
2626  4500000.0         5       5.50         6640     40014     2.0   
3914  7062500.0         5       4.50        10040     37325     2.0   
4411  5570000.0         5       5.75         9200     35069     2.0   
7252  7700000.0         6       8.00        12050     27600     2.5   
8092  4668000.0         5       6.75         9640     13068     1.0   
8638  4489000.0         4       3.00         6430     27517     2.0   
9254  6885000.0         6       7.75         9890     31374     2.0   

      waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
1164           1     4          3     12        5990           2020      1999   
1315           1     4          4     12        5000           2390      1991   
1448           0     4          3     12        6720           1280      2009   
2626           1     4          3     12        6350            290      2004   
3914           1     2          3     11        7680           2360      1940   
4411           0     0          3     13        6200           3000      2001   
7252           0     3          4     13        8570           3480      1910   
8092           1     4          3     12        4820           4820      1983   
8638           0     0          3     12        6430              0      2001   
9254           0     4          3     13        8860           1030      2001   

      yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
1164             0    98033  47.6767 -122.211           3430       26788  
1315             0    98040  47.5631 -122.210           4320       24619  
1448             0    98004  47.6232 -122.220           4600       21750  
2626             0    98155  47.7493 -122.280           3030       23408  
3914          2001    98004  47.6500 -122.214           3930       25449  
4411             0    98039  47.6289 -122.233           3560       24345  
7252          1987    98102  47.6298 -122.323           3940        8800  
8092          2009    98040  47.5570 -122.210           3270       10454  
8638             0    98004  47.6208 -122.219           3720       14592  
9254             0    98039  47.6305 -122.240           4540       42730

In [83]:
data[z_score_df.bedrooms>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
4096    599999.0         9       4.50         3830      6988     2.5   
4235    700000.0         9       3.00         3680      4400     2.0   
6079   1280000.0         9       4.50         3650      5000     2.0   
8546    450000.0         9       7.50         4050      6504     2.0   
8757    520000.0        11       3.00         3000      4960     2.0   
13314  1148000.0        10       5.25         4590     10920     1.0   
15161   650000.0        10       2.00         3610     11914     2.0   
15870   640000.0        33       1.75         1620      6000     1.0   
16844  1400000.0         9       4.00         4620      5508     2.5   
18443   934000.0         9       3.00         2820      4480     2.0   
19254   660000.0        10       3.00         2920      3745     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
4096            0     0          3      7        2450           1380   
4235            0     0          3      7        2830            850   
6079            0     0          3      8        2530           1120   
8546            0     0          3      7        4050              0   
8757            0     0          3      7        2400            600   
13314           0     2          3      9        2500           2090   
15161           0     0          4      7        3010            600   
15870           0     0          5      7        1040            580   
16844           0     0          3     11        3870            750   
18443           0     0          3      7        1880            940   
19254           0     0          4      7        1860           1060   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
4096       1938             0    98103  47.6927 -122.338           1460   
4235       1908             0    98102  47.6374 -122.324           1960   
6079       1915          2010    98105  47.6604 -122.289           2510   
8546       1996             0    98144  47.5923 -122.301           1448   
8757       1918          1999    98106  47.5560 -122.363           1420   
13314      2008             0    98004  47.5861 -122.113           2730   
15161      1958             0    98006  47.5705 -122.175           2040   
15870      1947             0    98103  47.6878 -122.331           1330   
16844      1915             0    98105  47.6684 -122.309           2710   
18443      1918             0    98105  47.6654 -122.307           2460   
19254      1913             0    98105  47.6635 -122.320           1810   

       sqft_lot15  
4096         6291  
4235         2450  
6079         5000  
8546         3866  
8757         4960  
13314       10400  
15161       11914  
15870        4700  
16844        4320  
18443        4400  
19254        3745

In [84]:
data[z_score_df.bathrooms>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
1315   5300000.0         6       6.00         7390     24829     2.0   
4024    800000.0         7       6.75         7480     41664     2.0   
4035   2150000.0         8       6.00         4340      9415     2.0   
7252   7700000.0         6       8.00        12050     27600     2.5   
8092   4668000.0         5       6.75         9640     13068     1.0   
8546    450000.0         9       7.50         4050      6504     2.0   
9254   6885000.0         6       7.75         9890     31374     2.0   
12370  4208000.0         5       6.00         7440     21540     2.0   
12777  2280000.0         7       8.00        13540    307752     3.0   
14556  2888000.0         5       6.25         8670     64033     2.0   
18302  3300000.0         5       6.25         8020     21738     2.0   
19148  3635000.0         5       6.00         5490     19897     2.0   
20578  1180000.0         6       6.50         6260     10955     2.0   
21050   900000.0         5       6.00         7120     40806     2.0   
21344  1488000.0         5       6.00         6880    279968     2.0   
21506  2238890.0         5       6.50         7270    130017     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
1315            1     4          4     12        5000           2390   
4024            0     2          3     11        5080           2400   
4035            0     0          3      8        4340              0   
7252            0     3          4     13        8570           3480   
8092            1     4          3     12        4820           4820   
8546            0     0          3      7        4050              0   
9254            0     4          3     13        8860           1030   
12370           0     0          3     12        5550           1890   
12777           0     4          3     12        9410           4130   
14556           0     4          3     13        6120           2550   
18302           0     0          3     11        8020              0   
19148           0     0          3     12        5490              0   
20578           0     0          3     11        4840           1420   
21050           0     4          3     12        5480           1640   
21344           0     3          3     12        4070           2810   
21506           0     0          3     12        6420            850   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
1315       1991             0    98040  47.5631 -122.210           4320   
4024       1953             0    98166  47.4643 -122.368           2810   
4035       1967             0    98004  47.6316 -122.202           2050   
7252       1910          1987    98102  47.6298 -122.323           3940   
8092       1983          2009    98040  47.5570 -122.210           3270   
8546       1996             0    98144  47.5923 -122.301           1448   
9254       2001             0    98039  47.6305 -122.240           4540   
12370      2003             0    98006  47.5692 -122.189           4740   
12777      1999             0    98053  47.6675 -121.986           4850   
14556      1965          2003    98177  47.7295 -122.372           4140   
18302      2001             0    98006  47.5675 -122.189           4160   
19148      2005             0    98039  47.6165 -122.236           2910   
20578      2007             0    98075  47.5947 -122.039           2710   
21050      2007             0    98006  47.5451 -122.114           3440   
21344      2007             0    98045  47.4624 -121.779           4690   
21506      2010             0    98027  47.5371 -121.982           1800   

       sqft_lot15  
1315        24619  
4024        33190  
4035         9100  
7252         8800  
8092        10454  
8546         3866  
9254        42730  
12370       19329  
12777      217800  
14556       81021  
18302       18969  
19148       17600  
20578       12550  
21050

In [85]:
data[z_score_df.sqft_living>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
1164   5110800.0         5       5.25         8010     45517     2.0   
1315   5300000.0         6       6.00         7390     24829     2.0   
1448   5350000.0         5       5.00         8000     23985     2.0   
2444   3278000.0         2       1.75         6840     10000     2.5   
2713   1110000.0         5       3.50         7350     12231     2.0   
3020   2525000.0         4       5.50         6930     45100     1.0   
3914   7062500.0         5       4.50        10040     37325     2.0   
4024    800000.0         7       6.75         7480     41664     2.0   
4149   4000000.0         4       5.50         7080     16573     2.0   
4411   5570000.0         5       5.75         9200     35069     2.0   
4811   2479000.0         5       3.75         6810      7500     2.5   
6501   1940000.0         4       5.75         7220    223462     2.0   
7035   3800000.0         5       5.50         7050     42840     1.0   
7252   7700000.0         6       8.00        12050     27600     2.5   
7907   3200000.0         6       5.00         7100     18200     2.5   
8092   4668000.0         5       6.75         9640     13068     1.0   
9254   6885000.0         6       7.75         9890     31374     2.0   
10373  2983000.0         5       5.50         7400     18898     2.0   
11685  1135000.0         6       4.25         6900    244716     2.0   
11871  1950000.0         4       3.25         7420    167869     2.0   
12370  4208000.0         5       6.00         7440     21540     2.0   
12777  2280000.0         7       8.00        13540    307752     3.0   
12872  3100000.0         6       4.25         6980     15682     3.0   
13411  2415000.0         5       4.75         7880     24250     2.0   
14032  1680000.0         4       3.75         7620     29536     2.0   
14556  2888000.0         5       6.25         8670     64033     2.0   
15482  3200000.0         4       3.25         7000     28206     1.0   
16773  1920000.0         5       5.75         7730    230868     2.0   
18302  3300000.0         5       6.25         8020     21738     2.0   
18477  3300000.0         8       4.00         7710     11750     3.5   
18594  1135250.0         5       4.00         7320    217800     2.0   
19858  2700000.0         4       4.00         7850     89651     2.0   
21050   900000.0         5       6.00         7120     40806     2.0   
21344  1488000.0         5       6.00         6880    279968     2.0   
21506  2238890.0         5       6.50         7270    130017     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
1164            1     4          3     12        5990           2020   
1315            1     4          4     12        5000           2390   
1448            0     4          3     12        6720           1280   
2444            1     4          3     11        4350           2490   
2713            0     4          3     11        4750           2600   
3020            0     0          4     11        4310           2620   
3914            1     2          3     11        7680           2360   
4024            0     2          3     11        5080           2400   
4149            0     0          3     12        5760           1320   
4411            0     0          3     13        6200           3000   
4811            0     0          3     13        6110            700   
6501            0     4          3     12        6220           1000   
7035            0     2          4     13        4320           2730   
7252            0     3          4     13        8570           3480   
7907            0     0          3     13        5240           1860   
8092            1     4          3     12        4820           4820   
9254            0     4          3     13        8860           1030   
10373           0     3          3     13        6290           1110   
11685           0     0          4      9        4

In [86]:
data[z_score_df.sqft_lot>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
145     921500.0         4       2.50         3670    315374     2.0   
527    1600000.0         6       5.00         6050    230652     2.0   
929     390000.0         4       3.00         2570    262018     1.0   
1045    590000.0         3       1.75         1560    242629     1.0   
1703    617000.0         3       1.75         3020    360241     2.0   
...          ...       ...        ...          ...       ...     ...   
20811   580000.0         3       2.50         1820    374616     2.0   
20974   950000.0         3       2.50         2780    275033     1.0   
21325   659000.0         3       2.50         3090    384634     2.0   
21344  1488000.0         5       6.00         6880    279968     2.0   
21351   565000.0         2       1.75         1130    276170     1.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
145             0     0          4      9        3670              0   
527             0     3          3     11        6050              0   
929             0     0          3      7        1420           1150   
1045            0     0          3      7        1560              0   
1703            0     0          3      8        3020              0   
...           ...   ...        ...    ...         ...            ...   
20811           0     0          3      7        1820              0   
20974           0     0          3     10        2780              0   
21325           0     0          3      8        3090              0   
21344           0     3          3     12        4070           2810   
21351           0     0          3      8        1130              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
145        1994             0    98077  47.7421 -122.026           2840   
527        2001             0    98024  47.6033 -121.943           4210   
929        1988             0    98058  47.4417 -122.090           2260   
1045       1981             0    98053  47.6493 -121.956           2320   
1703       1992             0    98092  47.2662 -122.088           1890   
...         ...           ...      ...      ...      ...            ...   
20811      1999             0    98014  47.6539 -121.959           1870   
20974      2006             0    98045  47.4496 -121.766           1680   
21325      2007             0    98019  47.7072 -121.927           2200   
21344      2007             0    98045  47.4624 -121.779           4690   
21351      2006             0    98022  47.2673 -122.027           2092   

       sqft_lot15  
145         87991  
527        233971  
929         19811  
1045       220654  
1703       209959  
...           ...  
20811      220654  
20974       16340  
21325      292645  
21344      256803  
21351      217800  

[148 rows x 19 columns]

In [87]:
data[z_score_df.sqft_lot>10]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
1719    700000.0         4       1.00         1300   1651359     1.0   
2964    999000.0         3       2.75         2830    505166     1.0   
3949    998000.0         4       3.25         3770    982998     2.0   
4387    480000.0         4       3.50         3370    435600     2.0   
4441    790000.0         2       3.00         2560    982278     1.0   
4540    550000.0         3       2.00         3650    843309     2.0   
5073    859000.0         3       2.50         2920    434728     2.0   
6691   1998000.0         2       2.50         3900    920423     2.0   
7077   1650000.0         4       3.25         3920    881654     3.0   
7250    950000.0         4       3.00         3230    438213     2.0   
7294    700000.0         2       1.75         1679    577605     2.0   
7647    542500.0         5       3.25         3010   1074218     1.5   
7769    855000.0         4       3.50         4030   1024068     2.0   
8443   2140000.0         4       3.75         5150    453895     2.0   
9210    350000.0         4       2.25         1570    499571     1.0   
9714    937500.0         4       4.00         5545    871200     2.0   
12919   428000.0         3       1.75         1580    507038     1.0   
13006   750000.0         3       2.50         2350    715690     1.5   
13250  1075000.0         3       3.00         4020    435600     1.5   
13464   790000.0         3       2.50         2640    432036     1.5   
13477   849900.0         2       2.00         2280    641203     2.0   
13886   575000.0         3       2.00         2690    435600     2.0   
14688   569950.0         4       3.50         2700    443440     1.5   
15934   745000.0         3       2.50         2810    435600     2.0   
16188   700000.0         3       2.50         2530    623779     1.0   
16995   667000.0         3       1.75         3320    478288     1.5   
17319   190000.0         2       1.00          710   1164794     1.0   
17349   800000.0         5       1.75         1930    501376     2.0   
17576   375000.0         1       1.00          800    533610     1.5   
17825   425000.0         3       2.75         1360    542322     1.0   
18227   525000.0         3       2.25         2040    435600     2.0   
18842   330000.0         2       2.00         1550    435600     1.5   
19091   745000.0         5       4.00         4720    493534     2.0   
20421   920000.0         4       3.75         4030    503989     2.0   
20452  1600000.0         4       5.50         6530    871200     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
1719            0     3          4      6        1300              0   
2964            1     3          4      8        1830           1000   
3949            0     0          3     10        3770              0   
4387            0     3          3      9        3370              0   
4441            0     0          3      8        2560              0   
4540            0     0          4      7        3650              0   
5073            0     3          4      8        2920              0   
6691            0     0          3     12        3900              0   
7077            0     3          3     11        3920              0   
7250            0     0          3      9        3230              0   
7294            0     0          3      9        1679              0   
7647            0     0          5      8        2010           1000   
7769            0     0          3     10        4030              0   
8443            0     3          3     11        4360            790   
9210            0     3          4      7        1570              0   
9714            0     0          3     11        3605           1940   
12919           0     2          4      7        1580              0   
13006           0     0          4      9        2350              0   
13250           0     2          3     10        4

In [88]:
data[z_score_df.floors>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [89]:
data[z_score_df.waterfront>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
49     1350000.0         3       2.50         2753     65005     1.0   
230     655000.0         2       1.75         1450     15798     2.0   
246    2400000.0         4       2.50         3650      8354     1.0   
264     369900.0         1       0.75          760     10079     1.0   
300    3075000.0         4       5.00         4550     18641     1.0   
...          ...       ...        ...          ...       ...     ...   
19984  1898000.0         3       2.50         2830      4334     3.0   
20325  3000000.0         3       3.50         4410     10756     2.0   
20767  2300000.0         4       4.00         4360      8175     2.5   
21201  2230000.0         3       3.50         3760      5634     2.0   
21576  3567000.0         5       4.50         4850     10584     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
49              1     2          5      9        2165            588   
230             1     4          3      7        1230            220   
246             1     4          3      9        1830           1820   
264             1     4          5      5         760              0   
300             1     4          3     10        2600           1950   
...           ...   ...        ...    ...         ...            ...   
19984           1     4          3     10        2830              0   
20325           1     4          3     11        3430            980   
20767           1     4          3     10        3940            420   
21201           1     4          3     11        2830            930   
21576           1     4          3     10        3540           1310   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
49         1953             0    98070  47.4041 -122.451           2680   
230        1915          1978    98166  47.4497 -122.375           2030   
246        2000             0    98074  47.6338 -122.072           3120   
264        1936             0    98070  47.4683 -122.438           1230   
300        2002             0    98074  47.6053 -122.077           4550   
...         ...           ...      ...      ...      ...            ...   
19984      2006             0    98074  47.6318 -122.071           2830   
20325      2014             0    98056  47.5283 -122.205           3550   
20767      2007             0    98008  47.5724 -122.104           2670   
21201      2014             0    98056  47.5285 -122.205           3560   
21576      2007             0    98008  47.5943 -122.110           3470   

       sqft_lot15  
49          72513  
230         13193  
246         18841  
264         14267  
300         19508  
...           ...  
19984       38211  
20325        5634  
20767        8525  
21201        5762  
21576       18270  

[163 rows x 19 columns]

In [90]:
data[z_score_df.view>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [91]:
data[z_score_df.condition>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [92]:
data[z_score_df.grade>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
19452  142000.0         0        0.0          290     20875     1.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
19452           0     0          1      1         290              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
19452      1963             0    98024  47.5308 -121.888           1620   

       sqft_lot15  
19452       22850

In [93]:
data[z_score_df.sqft_above>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
419    1550000.0         5       4.25         6070    171626     2.0   
527    1600000.0         6       5.00         6050    230652     2.0   
1100   1570000.0         5       4.50         6070     14731     2.0   
1164   5110800.0         5       5.25         8010     45517     2.0   
1448   5350000.0         5       5.00         8000     23985     2.0   
2626   4500000.0         5       5.50         6640     40014     2.0   
3121   1320000.0         4       5.25         6110     10369     2.0   
3914   7062500.0         5       4.50        10040     37325     2.0   
4411   5570000.0         5       5.75         9200     35069     2.0   
4811   2479000.0         5       3.75         6810      7500     2.5   
5059   1525000.0         4       3.50         5990    111078     2.0   
6501   1940000.0         4       5.75         7220    223462     2.0   
7252   7700000.0         6       8.00        12050     27600     2.5   
8638   4489000.0         4       3.00         6430     27517     2.0   
9254   6885000.0         6       7.75         9890     31374     2.0   
10373  2983000.0         5       5.50         7400     18898     2.0   
10951  1610000.0         5       4.50         6085    142725     3.0   
11105  1820000.0         4       4.50         6640     53330     2.0   
11871  1950000.0         4       3.25         7420    167869     2.0   
12777  2280000.0         7       8.00        13540    307752     3.0   
13400  1822500.0         4       4.50         6380     88714     2.0   
13411  2415000.0         5       4.75         7880     24250     2.0   
14032  1680000.0         4       3.75         7620     29536     2.0   
14556  2888000.0         5       6.25         8670     64033     2.0   
16773  1920000.0         5       5.75         7730    230868     2.0   
18302  3300000.0         5       6.25         8020     21738     2.0   
18477  3300000.0         8       4.00         7710     11750     3.5   
18594  1135250.0         5       4.00         7320    217800     2.0   
19858  2700000.0         4       4.00         7850     89651     2.0   
20452  1600000.0         4       5.50         6530    871200     2.0   
20471  1800000.0         4       3.50         6370    205603     2.0   
21506  2238890.0         5       6.50         7270    130017     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
419             0     0          3     12        6070              0   
527             0     3          3     11        6050              0   
1100            0     0          3     11        6070              0   
1164            1     4          3     12        5990           2020   
1448            0     4          3     12        6720           1280   
2626            1     4          3     12        6350            290   
3121            0     0          3     11        6110              0   
3914            1     2          3     11        7680           2360   
4411            0     0          3     13        6200           3000   
4811            0     0          3     13        6110            700   
5059            0     0          3     11        5990              0   
6501            0     4          3     12        6220           1000   
7252            0     3          4     13        8570           3480   
8638            0     0          3     12        6430              0   
9254            0     4          3     13        8860           1030   
10373           0     3          3     13        6290           1110   
10951           0     0          3     11        6085              0   
11105           0     0          3     12        6640              0   
11871           0     3          3     12        7420              0   
12777           0     4          3     12        9410           4130   
13400           0     0          3     12        6380              0   
13411           0     2          3     13        7

In [94]:
data[z_score_df.sqft_basement>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
2088   1195000.0         5       3.25         5180     19606     1.0   
2125   1325000.0         3       3.75         6400     76665     1.0   
2713   1110000.0         5       3.50         7350     12231     2.0   
3020   2525000.0         4       5.50         6930     45100     1.0   
4411   5570000.0         5       5.75         9200     35069     2.0   
5049   1385000.0         6       2.75         5700     20000     1.0   
6628    850000.0         4       2.75         5440    239580     1.0   
7035   3800000.0         5       5.50         7050     42840     1.0   
7252   7700000.0         6       8.00        12050     27600     2.5   
8092   4668000.0         5       6.75         9640     13068     1.0   
10085  1900000.0         5       4.25         6510     16471     2.0   
11829  1785000.0         4       3.50         5490     14300     1.0   
12777  2280000.0         7       8.00        13540    307752     3.0   
14556  2888000.0         5       6.25         8670     64033     2.0   
15049  1535000.0         4       4.50         5770     10050     1.0   
15482  3200000.0         4       3.25         7000     28206     1.0   
19215  2000000.0         5       4.25         6490     10862     2.0   
21344  1488000.0         5       6.00         6880    279968     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
2088            0     0          3     11        2610           2570   
2125            0     2          4     10        3810           2590   
2713            0     4          3     11        4750           2600   
3020            0     0          4     11        4310           2620   
4411            0     0          3     13        6200           3000   
5049            0     4          4     10        2850           2850   
6628            0     0          2      9        2720           2720   
7035            0     2          4     13        4320           2730   
7252            0     3          4     13        8570           3480   
8092            1     4          3     12        4820           4820   
10085           0     3          4     11        3250           3260   
11829           0     4          3     12        2910           2580   
12777           0     4          3     12        9410           4130   
14556           0     4          3     13        6120           2550   
15049           0     3          5      9        3160           2610   
15482           1     4          4     12        3500           3500   
19215           0     3          4     11        3940           2550   
21344           0     3          3     12        4070           2810   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
2088       1993             0    98006  47.5550 -122.114           4050   
2125       1966             0    98177  47.7313 -122.370           3430   
2713       2001             0    98065  47.5373 -121.865           5380   
3020       1950          1991    98006  47.5547 -122.144           2560   
4411       2001             0    98039  47.6289 -122.233           3560   
5049       1977             0    98006  47.5601 -122.160           3690   
6628       1969             0    98001  47.3540 -122.293           1970   
7035       1978             0    98004  47.6229 -122.220           5070   
7252       1910          1987    98102  47.6298 -122.323           3940   
8092       1983          2009    98040  47.5570 -122.210           3270   
10085      1980             0    98040  47.5758 -122.242           4480   
11829      1996             0    98006  47.5511 -122.114           4290   
12777      1999             0    98053  47.6675 -121.986           4850   
14556      1965          2003    98177  47.7295 -122.372           4140   
15049      1949             0    98115  47.6770 -122.275           2950   
15482      1991             0    98075  47.5928 -122.086           4913  

In [95]:
data[z_score_df.yr_built>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [96]:
data[z_score_df.yr_renovated>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [97]:
data[z_score_df.zipcode>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [98]:
data[z_score_df.lat>5]

Empty DataFrame
Columns: [price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

In [99]:
data[z_score_df.long>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
2589   134000.0         2       1.50          980      5000     2.0   
2927   167000.0         1       1.00          780     10235     1.5   
4203   150000.0         3       0.75          490     38500     1.5   
4848   525000.0         3       2.75         2100     10362     2.0   
5867   175000.0         2       1.75         1050      9800     1.5   
6089   150000.0         3       1.00          890      6488     1.5   
10095  200000.0         2       1.75         1320     13052     1.5   
10898  241000.0         2       1.75         1070      9750     1.5   
13072  155000.0         2       1.00         1010     43056     1.5   
13249  375000.0         3       1.75         2140     13598     1.5   
13994  370000.0         2       2.00         2100     58488     2.0   
14615  160000.0         3       1.00         1340     18552     1.5   
16842  170000.0         1       0.75          850      5600     1.0   
16941  160000.0         3       1.00         1140      3240     1.5   
19650  155000.0         3       1.00         1300      6098     1.0   
19980  415000.0         3       2.75         2600     64626     1.5   
21386  320000.0         3       1.75         1790     66250     1.5   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
2589            0     0          3      7         980              0   
2927            0     0          3      6         780              0   
4203            0     0          4      5         490              0   
4848            0     0          3      9        1510            590   
5867            0     0          4      6        1050              0   
6089            0     0          3      5         890              0   
10095           0     0          3      7        1320              0   
10898           0     0          3      7        1070              0   
13072           0     0          3      5        1010              0   
13249           0     0          4      7        1620            520   
13994           0     0          3      9        2100              0   
14615           0     0          4      5        1340              0   
16842           0     2          3      6         850              0   
16941           0     0          4      6        1140              0   
19650           0     0          3      7        1300              0   
19980           0     0          3      8        2600              0   
21386           0     0          3      7        1790              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
2589       1922          2003    98014  47.7076 -121.359           1040   
2927       1989             0    98014  47.7130 -121.315            930   
4203       1959             0    98014  47.7112 -121.315            800   
4848       1998             0    98045  47.4347 -121.417           2240   
5867       1975             0    98019  47.7595 -121.473           1230   
6089       1928             0    98014  47.7087 -121.352           1330   
10095      1980             0    98014  47.7120 -121.352           1320   
10898      1995             0    98014  47.7131 -121.319            970   
13072      1990             0    98014  47.7105 -121.316            830   
13249      1970             0    98014  47.7139 -121.321            930   
13994      2005             0    98014  47.7200 -121.402           1440   
14615      1935             0    98014  47.7129 -121.325            960   
16842      1903          1994    98019  47.7654 -121.480            900   
16941      1910             0    98014  47.7093 -121.364           1140   
19650      2013             0    98014  47.7074 -121.364           1300   
19980      2009             0    98014  47.7185 -121.405           1740   
21386      2003             0    98014  47.7179 -121.403           1440   

       sqft_lot15  
2589         5000  
2927        10165  
4203        18297  
484

In [100]:
data[z_score_df.sqft_living15>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
1123   1200000.0         4       1.75         3990     13470     2.0   
1530   1250000.0         4       3.75         3830     41263     2.0   
5451   1780000.0         4       3.25         4890     13402     2.0   
10373  2983000.0         5       5.50         7400     18898     2.0   
11871  1950000.0         4       3.25         7420    167869     2.0   
12713  2408000.0         5       2.50         4600     23250     1.5   
16430  1750000.0         6       4.25         5860     13928     2.0   
19858  2700000.0         4       4.00         7850     89651     2.0   
20563  1240420.0         5       3.25         5790     13726     2.0   
20830  1750000.0         5       3.25         5790     12739     2.0   
21540  1881580.0         5       3.50         4410     13000     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
1123            0     0          3     11        3990              0   
1530            0     0          4     11        3830              0   
5451            0     0          3     13        4890              0   
10373           0     3          3     13        6290           1110   
11871           0     3          3     12        7420              0   
12713           0     2          3      9        3600           1000   
16430           0     3          3     10        4150           1710   
19858           0     0          3     12        7850              0   
20563           0     3          3     10        4430           1360   
20830           0     3          3     10        4430           1360   
21540           0     3          3     10        2920           1490   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
1123       2006             0    98059  47.5305 -122.131           5790   
1530       1990             0    98077  47.7237 -122.042           5600   
5451       2004             0    98059  47.5303 -122.131           5790   
10373      2001             0    98006  47.5431 -122.112           6110   
11871      2002             0    98045  47.4548 -121.764           5610   
12713      1918          2003    98004  47.6230 -122.218           5500   
16430      2013             0    98006  47.5382 -122.114           5790   
19858      2006             0    98027  47.5406 -121.982           6210   
20563      2014             0    98006  47.5388 -122.114           5790   
20830      2014             0    98006  47.5380 -122.114           5790   
21540      2014             0    98006  47.5382 -122.111           5790   

       sqft_lot15  
1123        13709  
1530        56568  
5451        13539  
10373       26442  
11871      169549  
12713       20066  
16430       13928  
19858       95832  
20563       13726  
20830       13928  
21540       12969

In [101]:
data[z_score_df.sqft_lot15>5]

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
147     430000.0         2       2.50         2420     60984     2.0   
416     563500.0         4       1.75         2085    174240     1.0   
419    1550000.0         5       4.25         6070    171626     2.0   
443     350000.0         3       1.50         1250    219978     1.0   
484    1385000.0         4       3.25         4860    181319     2.5   
...          ...       ...        ...          ...       ...     ...   
21325   659000.0         3       2.50         3090    384634     2.0   
21344  1488000.0         5       6.00         6880    279968     2.0   
21351   565000.0         2       1.75         1130    276170     1.0   
21431   800000.0         4       3.25         3540    159430     2.0   
21548  1057000.0         2       1.50         2370    184231     2.0   

       waterfront  view  condition  grade  sqft_above  sqft_basement  \
147             0     0          3      7        2420              0   
416             0     0          3      7        1610            475   
419             0     0          3     12        6070              0   
443             0     0          4      6        1250              0   
484             0     0          3      9        4860              0   
...           ...   ...        ...    ...         ...            ...   
21325           0     0          3      8        3090              0   
21344           0     3          3     12        4070           2810   
21351           0     0          3      8        1130              0   
21431           0     0          3      9        3540              0   
21548           0     0          3     11        2370              0   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
147        2007             0    98027  47.5262 -121.943           1940   
416        1964             0    98024  47.5753 -121.950           2690   
419        1999             0    98024  47.5954 -121.950           4680   
443        1980             0    98038  47.4056 -121.955           1930   
484        1993             0    98074  47.6179 -122.005           3850   
...         ...           ...      ...      ...      ...            ...   
21325      2007             0    98019  47.7072 -121.927           2200   
21344      2007             0    98045  47.4624 -121.779           4690   
21351      2006             0    98022  47.2673 -122.027           2092   
21431      2007             0    98014  47.6285 -121.899           1940   
21548      2005             0    98045  47.4543 -121.778           3860   

       sqft_lot15  
147        193842  
416        174240  
419        211267  
443        210394  
484        181319  
...           ...  
21325      292645  
21344      256803  
21351      217800  
21431      392040  
21548      151081  

[227 rows x 19 columns]

# Summary and Decision on removal or modification

1) Feature with obvious wrong or self-contradictory data
bedrooms: one outlier has 33 bedrooms but with only 1620 sqft_living, which is impossible. This record (#15870) should be removed. Other outliers are 9-11, reasonable when compared with the size, possibly multi-family homes.

2) Features with no outliers detected by threshold in the above: floors, view, condition, grade, yr_built, yr_renovated, zipcode, lat

Among these above features, further checking:

floors: whole range within 1-3.5, looks sensible.

view, condition and grade: all fall within defined range

yr_built: oldest is 1900, not usual but possible

yr_renovated: 
oldest is 1934 - looks odd, but more important issue is its overall data quality alerted by sparse data - only 5% is non-zero, so on average a house is renovated only every 20 years, perhaps still reasonable overall, but with some very old years of renovation, and subjective definition of major renovation, instead of taking out a few records, we may need to consider whether to use this feature (to be examined during further checking or training). One alternative is to consider modification by replacing zero with the value of yr_built, as this year should reflect the condition of those properties truly without renovation. Also, another problem is whether this feature (if of good data quality) would be redundant given that there is already a feature "condition".

zipcode: confirmed with list of zipcodes of King County

lat: using Google Map, the range of King County should be 47.08 to 47.78, the data is consistent with this range

3) Features with outliers but can be explained trivially

long: using Google Map, the range of King County shoud be -122.53 to -121.06, the data is consistent with this range

waterfront: 
all outliers are labelled 1, and actually it is defined as either 0 or 1 with only 0.8% labelled as 1, so this is data sparsity or potentially data quality issue instead. Depending on the landscape the data may be genuinely sparse. Instead of removing all these outliers, we should consider whether this feature should be used (to be examined during further checking or training), as there is also another issue - whether it is redundant given that there is already a feature "view".

bathrooms: outliers are at 6-8, reasonable when compared with size (at least 4050 sqft_living)

4) Features about size and price

Sqft_living: note that it is simply the sum of sqft_above and sqft_basement - so one of these feature columns should be removed for data training. Among these 3 features we focus on sqft_living, which can be up to 13540 - this is rare but possible as the largest house for sale in Seattle is around this size.

Sqft_living15: similar to sqft_living but it refers to nearest 15 neighbours'average sqft_living. Outliers up to 6210. By similar argument as sqft_living, it is rare but possible. 

Sqft_lot: outliers up to 1651359 , a very large number but it can be the case for farmhouse with large land lot. We give the benefit of doubt to data as this feature is hard to judge correctness by itself. 

Sqft_lot15: similar to sqft_lot, we give benefit of doubt to data.

Price: with second screening using threshold at z score of 10, outliers are at 4.5m - 7.7m, all with sizeable sqft_living, and these amounts are in par with the most luxuary homes in Seattle. 

